File: theo_ha/5/Templates/C/Python/regex.py
from abc import ABC, abstractmethod
import sys


class Visitor(ABC):
    @abstractmethod
    def visit_empty(self):
        pass

    @abstractmethod
    def visit_epsilon(self):
        pass

    @abstractmethod
    def visit_single(self, c):
        pass

    @abstractmethod
    def visit_alternative(self, rs):
        pass

    @abstractmethod
    def visit_concat(self, rs):
        pass

    @abstractmethod
    def visit_star(self, r):
        pass


class Regex(ABC):

    single_instances = {}

    def __init__(self):
        pass

    def to_str(self, precedence):
        if precedence > self.get_precedence():
            return "({})".format(str(self))
        return str(self)

    @abstractmethod
    def get_precedence(self):
        pass

    @abstractmethod
    def __str__(self):
        pass

    @abstractmethod
    def is_nullable(self):
        pass

    # Smart constructors
    @staticmethod
    def empty():
        return Empty()

    @staticmethod
    def epsilon():
        return Epsilon()

    @staticmethod
    def single(c):
        if c in Regex.single_instances:
            return Regex.single_instances[c]
        r = Single(c)
        Regex.single_instances[c] = r
        return r

    @staticmethod
    def alternatives(regexes, *args):
        if len(args) > 0:
            regexes = [regexes] + list(args)
        s = set()
        nullable = False
        has_epsilon = False

        for r in regexes:
            if isinstance(r, Epsilon):
                has_epsilon = True
            else:
                nullable = nullable or r.is_nullable()
                if isinstance(r, Alternative):
                    for r2 in r:
                        s.add(r)
                elif not isinstance(r, Empty):
                    s.add(r)

        if has_epsilon and not nullable:
            s.add(Regex.epsilon())

        if len(s) == 0:
            return Regex.empty()
        elif len(s) == 1:
            return s.pop()
        else:
            return Alternative(s)

    @staticmethod
    def concats(regexes, *args):
        if len(args) > 0:
            regexes = [regexes] + list(args)
        l = []
        for r in regexes:
            if isinstance(r, Empty):
                return Regex.empty()
            elif isinstance(r, Concat):
                for r2 in r:
                    l.append(r2)
            elif not isinstance(r, Epsilon):
                l.append(r)

        if len(l) == 0:
            return Regex.epsilon()
        elif len(l) == 1:
            return l[0]
        else:
            return Concat(l)

    @staticmethod
    def star(r):
        if isinstance(r, Star):
            return r
        elif isinstance(r, Empty) or isinstance(r, Epsilon):
            return Regex.epsilon()
        else:
            return Star(r)

    @abstractmethod
    def accept(self, vis):
        pass


class Single(Regex):

    def __init__(self, c):
        if len(c) != 1:
            raise ValueError("Character should be a string of length 1.")
        self.c = c
        self.size = 1

    def __str__(self):
        return self.c

    def __hash__(self):
        prime = 31
        return prime * 1 + ord(self.c)

    def __eq__(self, other):
        if not isinstance(other, Single):
            return False
        return self.c == other.c

    def get_precedence(self):
        return 4

    def is_nullable(self):
        return False

    def accept(self, vis):
        return vis.visit_single(self.c)


class Empty(Regex):

    def __init__(self):
        self.size = 1

    def __str__(self):
        return "{}"

    def __hash__(self):
        return 23

    def __self__(self, other):
        if not isinstance(other, Empty):
            return False
        return True

    def get_precedence(self):
        return 4

    def is_nullable(self):
        return False

    def accept(self, vis):
        return vis.visit_empty()


class Epsilon(Regex):

    def __init__(self):
        self.size = 1

    def __str__(self):
        return "()"

    def __hash__(self):
        return 47

    def __self__(self, other):
        if not isinstance(other, Epsilon):
            return False
        return True

    def get_precedence(self):
        return 4

    def is_nullable(self):
        return True

    def accept(self, vis):
        return vis.visit_epsilon()


class Alternative(Regex):

    def __init__(self, regexes):
        children = set()
        size = 1
        nullable = False
        for r in regexes:
            if isinstance(r, Empty):
                continue
            children.add(r)
            size += r.size
            nullable = nullable or r.is_nullable()
        self.children = children
        self.size = size
        self.nullable = nullable

    def __str__(self):
        return "|".join(r.to_str(self.get_precedence()) for r in self.children)

    def __iter__(self):
        return iter(self.children)

    def __eq__(self, other):
        if not isinstance(other, Alternative):
            return False
        return other.children == self.children

    def __hash__(self):
        prime = 31
        result = 1
        result = result * prime + \
            (hash(tuple(self.children)) if self.children else 0)
        result = result * prime + self.size
        return result

    def get_precedence(self):
        return 0

    def is_nullable(self):
        return self.nullable

    def accept(self, vis):
        return vis.visit_alternative(self.children)


class Concat(Regex):

    def __init__(self, regexes):
        children = list()
        size = 1
        nullable = True
        for r in regexes:
            if not isinstance(r, Epsilon):
                children.append(r)
                size += r.size
                nullable = nullable and r.is_nullable()
        self.size = size
        self.nullable = nullable
        self.children = children

    def __str__(self):
        return "".join(r.to_str(self.get_precedence()) for r in self.children)

    def __iter__(self):
        return iter(self.children)

    def __eq__(self, other):
        if not isinstance(other, Concat):
            return False
        return other.children == self.children

    def __hash__(self):
        prime = 31
        result = 1
        result = result * prime + \
            (hash(tuple(self.children)) if self.children else 0)
        result = result * prime + self.size
        return result

    def get_precedence(self):
        return 3

    def is_nullable(self):
        return self.nullable

    def accept(self, vis):
        return vis.visit_concat(self.children)


class Star(Regex):
    def __init__(self, r):
        self.r = r
        self.size = r.size + 1

    def __str__(self):
        s = self.r.to_str(self.get_precedence())
        return "{}*".format(s)

    def __eq__(self, other):
        if not isinstance(other, Star):
            return False
        return other.r == self.r

    def __hash__(self):
        prime = 31
        result = 1
        result = result * prime + (hash(self.r) if self.r else 0)
        result = result * prime + self.size
        return result

    def get_precedence(self):
        return 2

    def is_nullable(self):
        return True

    def accept(self, vis):
        return vis.visit_star(self.r)


def is_symbol(c: str) -> str:
    return c.isalnum()


class Regex_Parser:

    def __init__(self, s: str):
        self.input = s
        self.pos = 0

    def eof(self):
        return self.pos >= len(self.input)

    def peek(self):
        while not self.eof() and self.input[self.pos].isspace():
            self.pos += 1
        if self.eof():
            return None
        return self.input[self.pos]

    def pretty_peek(self) -> str:
        c = self.peek()
        if c:
            return "'{}'".format(c)
        return "end of input"

    def parse(self) -> Regex:
        r = self.parse_alternative()
        if not self.eof():
            raise ValueError(
                "Expected end of input, but got {} instead at position {}.".format(self.pretty_peek(), self.pos))

        return r

    def parse_alternative(self):
        r = self.parse_non_alternative()
        l = [r]
        c = self.peek()
        while c == '|':
            self.pos += 1
            l.append(self.parse_alternative())
            c = self.peek()

        if c != None and c != ')':
            raise ValueError(
                "Expected '|' or end of input, but got {} instead at position {}.".format(self.pretty_peek(), self.pos))

        return Regex.alternatives(l)

    def parse_non_alternative(self):
        c = self.peek()

        if c == None or c == ')' or c == '|':
            return Regex.epsilon()

        l = []
        r = None
        while True:
            if c == '{':
                self.pos += 1
                r = Regex.empty()
                if self.peek() != '}':
                    raise ValueError(
                        "Expected '}', but got {} instead at position {}.".format(self.pretty_peek(), self.pos))
            elif c == '∅':
                r = Regex.empty()
            elif c == 'ε' or c == '*':
                r = Regex.epsilon()
            elif c == '(':
                self.pos += 1
                r = self.parse_alternative()
                if (self.peek() != ')'):
                    raise ValueError("Expected ')', but got {} instead at position {}.".format(
                        self.pretty_peek(), self.pos))
            elif is_symbol(c):
                r = Regex.single(c)
            else:
                raise ValueError(
                    "Expected alphabet symbol or one of '(', '|', ')', '*', but got {} instead at position {}.".format(self.pretty_peek(), self.pos))

            self.pos += 1

            while(self.peek() == '*'):
                r = Regex.star(r)
                self.pos += 1

            l.append(r)

            c = self.peek()

            if c == None or c == ')' or c == '|':
                break

        return Regex.concats(l)


epsilon = Regex.epsilon
empty = Regex.empty
single = Regex.single
star = Regex.star
concats = Regex.concats
alternatives = Regex.alternatives

File: theo_ha/5_sol/Lösungen/A/Java/Regex.java

import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Stream;

public abstract class Regex {
	
	private static final class ReadOnlyIterator<T> implements Iterator<T> {

		private final Iterator<T> it;
		
		public ReadOnlyIterator(Iterator<T> it) {
			this.it = it;
		}
		
		@Override
		public boolean hasNext() {
			return it.hasNext();
		}

		@Override
		public T next() {
			return it.next();
		}
		
	}
	
	public static final class Empty extends Regex {
		
		@Override
		protected void appendTo(StringBuilder sb, int prec) {
	      sb.append("{}");
		}
		
		private Empty() {
		}
		
		@Override
		protected int getPrecedence() {
			return 4;
		}
		
		@Override
		public int hashCode() {
			return 23;
		}
		
		@Override
		public boolean equals(Object obj) {
			if (this == obj)
				return true;
			if (obj == null)
				return false;
			return (getClass() == obj.getClass());
		}
		
		public int size() {
			return 1;
		}

		@Override
		public boolean isNullable() {
			return false;
		}

		@Override
		public <R> R accept(Visitor<R> vis) {
			return vis.visitEmpty();
		}
		
	}
	
	public static final class Epsilon extends Regex {

		private Epsilon() {
		}
		
		@Override
		protected void appendTo(StringBuilder sb, int prec) {
	      sb.append("()");		
		}

		@Override
		protected int getPrecedence() {
			return 4;
		}
		
		@Override
		public int hashCode() {
			return 47;
		}
		
		@Override
		public boolean equals(Object obj) {
			if (this == obj)
				return true;
			if (obj == null)
				return false;
			return (getClass() == obj.getClass());
		}
		
		public int size() {
			return 1;
		}

		@Override
		public boolean isNullable() {
			return true;
		}
		
		@Override
		public <R> R accept(Visitor<R> vis) {
			return vis.visitEpsilon();
		}
		
	}

	public static final class Single extends Regex {
		
		private final char c;
		
		@Override
		protected int getPrecedence() {
			return 4;
		}
		
		public char getChar() {
			return c;
		}

		private Single(char c) {
			super();
			this.c = c;
		}
		
		@Override
		protected void appendTo(StringBuilder sb, int prec) {
			sb.append(c);
		}
		
		@Override
		public int hashCode() {
			final int prime = 31;
			int result = 1;
			result = prime * result + c;
			return result;
		}

		@Override
		public boolean equals(Object obj) {
			if (this == obj)
				return true;
			if (obj == null)
				return false;
			if (getClass() != obj.getClass())
				return false;
			Single other = (Single) obj;
			if (c != other.c)
				return false;
			return true;
		}
		
		public int size() {
			return 1;
		}

		@Override
		public boolean isNullable() {
			return false;
		}
		
		@Override
		public <R> R accept(Visitor<R> vis) {
			return vis.visitSingle(c);
		}
		
	}

	public static final class Concat extends Regex implements Iterable<Regex> {

		private final List<Regex> children;
		private final int size;
		private final boolean nullable;
		
		private Concat(List<Regex> children) {
			super();
			List<Regex> children2 = new LinkedList<>();
			int size = 1;
			boolean nullable = true;
			for (Regex r : children) {
				if (r instanceof Epsilon) continue;
				children2.add(r);
				size += r.size();
				nullable = nullable && r.isNullable();
			}
			this.size = size;
			this.nullable = nullable;
			this.children = Collections.unmodifiableList(children2);
		}
		
		public List<Regex> getChildren() {
			return children;
		}
		
		@Override
		protected void appendTo(StringBuilder sb, int prec) {
			for (Regex r : children)
				r.toStringAux(sb, getPrecedence());			
		}

		@Override
		public int hashCode() {
			final int prime = 31;
			int result = 1;
			result = prime * result + ((children == null) ? 0 : children.hashCode());
			result = prime * result + size;
			return result;
		}

		@Override
		public boolean equals(Object obj) {
			if (this == obj)
				return true;
			if (obj == null)
				return false;
			if (getClass() != obj.getClass())
				return false;
			Concat other = (Concat) obj;
			if (children == null) {
				if (other.children != null)
					return false;
			} else if (!children.equals(other.children))
				return false;
			if (size != other.size)
				return false;
			return true;
		}

		@Override
		protected int getPrecedence() {
			return 1;
		}
		
		public int size() {
			return size;
		}

		@Override
		public boolean isNullable() {
			return nullable;
		}

		@Override
		public Iterator<Regex> iterator() {
			return new ReadOnlyIterator<Regex>(children.iterator());
		}
		
		public Stream<Regex> stream() {
			return children.stream();
		}
		
		@Override
		public <R> R accept(Visitor<R> vis) {
			return vis.visitConcat(children);
		}
		
	}
	
	public static final class Alternative extends Regex implements Iterable<Regex> {
		
		private final Set<Regex> children;
		private final int size;
		private final boolean nullable;
		
		private Alternative(Iterable<Regex> children) {
			super();
			Set<Regex> children2 = new HashSet<Regex>();
			int size = 1;
			boolean nullable = false;
			for (Regex r : children) {
				if (r instanceof Empty) continue;
				children2.add(r);
				size += r.size();
				nullable = nullable || r.isNullable();
			}
			this.children = Collections.unmodifiableSet(children2);
			this.size = size;
			this.nullable = nullable;
		}
		
		public Set<Regex> getChildren() {
			return children;
		}
		
		@Override
		protected void appendTo(StringBuilder sb, int prec) {
			boolean first = true;
			for (Regex r : children) {
				if (!first) {
					sb.append("|");
				} else {
					first = false;
				}
				r.toStringAux(sb, getPrecedence());			
			}
		}

		@Override
		protected int getPrecedence() {
			return 0;
		}
		
		public int size() {
			return size;
		}

		@Override
		public int hashCode() {
			final int prime = 31;
			int result = 1;
			result = prime * result + ((children == null) ? 0 : children.hashCode());
			result = prime * result + size;
			return result;
		}

		@Override
		public boolean equals(Object obj) {
			if (this == obj)
				return true;
			if (obj == null)
				return false;
			if (getClass() != obj.getClass())
				return false;
			Alternative other = (Alternative) obj;
			if (children == null) {
				if (other.children != null)
					return false;
			} else if (!children.equals(other.children))
				return false;
			if (size != other.size)
				return false;
			return true;
		}

		@Override
		public boolean isNullable() {
			return nullable;
		}

		@Override
		public Iterator<Regex> iterator() {
			return new ReadOnlyIterator<Regex>(children.iterator());
		}
		
		public Stream<Regex> stream() {
			return children.stream();
		}
		
		@Override
		public <R> R accept(Visitor<R> vis) {
			return vis.visitAlternative(children);
		}
		
	}
	
	public static final class Star extends Regex {

		private final Regex r;
		private final int size;
		
		public Regex getRegex() {
			return r;
		}
		
		private Star(Regex r) {
			super();
			this.r = r;
			this.size = r.size() + 1;
		}
		
		@Override
		protected int getPrecedence() {
			return 2;
		}

		@Override
		protected void appendTo(StringBuilder sb, int prec) {
			r.toStringAux(sb, getPrecedence());
			sb.append('*');
		}

		@Override
		public int hashCode() {
			final int prime = 31;
			int result = 1;
			result = prime * result + ((r == null) ? 0 : r.hashCode());
			return result;
		}

		@Override
		public boolean equals(Object obj) {
			if (this == obj)
				return true;
			if (obj == null)
				return false;
			if (getClass() != obj.getClass())
				return false;
			Star other = (Star) obj;
			if (r == null) {
				if (other.r != null)
					return false;
			} else if (!r.equals(other.r))
				return false;
			return true;
		}
		
		public int size() {
			return size;
		}

		@Override
		public boolean isNullable() {
			return true;
		}
		
		@Override
		public <R> R accept(Visitor<R> vis) {
			return vis.visitStar(r);
		}
		
	}
	
	/**
	 * Visitor interface for regular expressions, with result type R. 
	 */
	public static interface Visitor<R> {
		public R visitEmpty();
		public R visitEpsilon();
		public R visitSingle(char c);
		public R visitAlternative(Set<Regex> rs);
		public R visitConcat(List<Regex> rs);
		public R visitStar(Regex r);
	}


	protected abstract void appendTo(StringBuilder sb, int prec);

	protected abstract int getPrecedence();
	
	public abstract <R> R accept(Visitor<R> vis);

	protected final void toStringAux(StringBuilder sb, int prec) {
		if (prec > getPrecedence()) {
			sb.append("(");
			appendTo(sb, getPrecedence());
			sb.append(")");
		} else {
			appendTo(sb, getPrecedence());
		}
	}

	public String toString() {
		StringBuilder sb = new StringBuilder();
		toStringAux(sb, 0);
		return sb.toString();
	}


	
	// Smart constructors
	private static final Empty EMPTY = new Empty();
	private static final Epsilon EPSILON = new Epsilon();
	private static final Map<Character, Single> singleInsts = new HashMap<>();
	
	public static final Regex empty() {
		return EMPTY;
	}

	public static final Regex epsilon() {
		return EPSILON;
	}
	
	public static final Regex single(char c) {
		Single r = singleInsts.get(c);
		if (r == null) {
			r = new Single(c);
			singleInsts.put(c, r);
		}
		return r;
	}

	public final Regex alternative(Iterator<Regex> it) {
		Set<Regex> s = new HashSet<Regex>();

		boolean nullable = false, hasEpsilon = false;
		Regex r = this;

		do {
			if (r instanceof Epsilon) {
				hasEpsilon = true;
			} else {
				nullable = nullable || r.isNullable();
				if (r instanceof Alternative) {
					for (Regex r2 : (Alternative) r)
						s.add(r2);
				} else {
					if (!(r instanceof Empty)) s.add(r);
				}
			}
			r = (it.hasNext()) ? it.next() : null;
		} while (r != null);
		
		if (hasEpsilon && !nullable)
			s.add(Regex.epsilon());

		if (s.size() == 0)
			return empty();
		else if (s.size() == 1)
			return s.iterator().next();
		else
			return new Alternative(s);
	}
	
	public final Regex alternative(Iterable<Regex> rs) {
		return alternative(rs.iterator());
	}

	public final Regex alternative(Stream<Regex> rs) {
		return alternative(rs.iterator());
	}
	
	public final Regex alternative(Regex... rs) {
		return alternative(Arrays.stream(rs));
	}
	
	public static final Regex alternatives(Iterator<Regex> it) {
		if (it.hasNext())
			return it.next().alternative(it);
		else
			return Regex.empty();
	}

	public static final Regex alternatives(Iterable<Regex> rs) {
		return alternatives(rs.iterator());
	}

	public static final Regex alternatives(Stream<Regex> rs) {
		return alternatives(rs.iterator());
	}
	
	public static final Regex alternatives(Regex... rs) {
		return alternatives(Arrays.stream(rs));
	}
	

	public final Regex concat(Iterator<Regex> it) {
		List<Regex> l = new LinkedList<Regex>();
		Regex r = this;
		do {
			if (r instanceof Empty) {
				return empty();
			} else if (r instanceof Concat) {
				for (Regex r2 : (Concat) r)
					l.add(r2);
			} else {
				if (!(r instanceof Epsilon)) l.add(r);
			}
			r = (it.hasNext()) ? it.next() : null;
		} while (r != null);

		if (l.size() == 0)
			return epsilon();
		else if (l.size() == 1)
			return l.iterator().next();
		else
			return new Concat(l);
	}
	
	public final Regex concat(Stream<Regex> rs) {
		return concat(rs.iterator());
	}
	
	public final Regex concat(Iterable<Regex> rs) {
		return concat(rs.iterator());
	}
	
	public final Regex concat(Regex... rs) {
		return concat(Arrays.stream(rs));
	}
	
	public static final Regex concats(Iterator<Regex> it) {
		if (it.hasNext())
			return it.next().concat(it);
		else
			return Regex.epsilon();
	}

	public static final Regex concats(Stream<Regex> rs) {
		return concats(rs.iterator());
	}
	
	public static final Regex concats(Iterable<Regex> rs) {
		return concats(rs.iterator());
	}
	
	public static final Regex concats(Regex... rs) {
		return concats(Arrays.stream(rs));
	}
	
	
	public static final Regex star(Regex r) {
		if (r instanceof Star)
			return r;
		else if (r instanceof Empty || r instanceof Epsilon)
			return epsilon();
		else
			return new Star(r);
	}
	
	public Regex star() {
		return Regex.star(this);
	}
	
	
	// Utility functions
	public abstract int size();

	/**
	 * Returns true iff the regular expression generates the empty word.
	 */
	public abstract boolean isNullable();
	
	public static final Regex parse(String s) {
		return (new RegexParser(s)).parse();
	}
	
	public Set<Character> getAlphabet() {
		final Set<Character> alphabet = new HashSet<Character>();
		var visitor = new Visitor<Void>() {
			public Void visitEmpty() {
				return null;
			}

			public Void visitEpsilon() {
				return null;
			}

			public Void visitSingle(char c) {
				alphabet.add(c);
				return null;
			}

			@Override
			public Void visitAlternative(Set<Regex> rs) {
				rs.stream().forEach(r -> r.accept(this));
				return null;
			}

			@Override
			public Void visitConcat(List<Regex> rs) {
				rs.stream().forEach(r -> r.accept(this));
				return null;
			}

			@Override
			public Void visitStar(Regex r) {
				r.accept(this);
				return null;
			}
		};
		accept(visitor);
		return alphabet;
	}
	
}
